ClimbSubsystem changes

galaxygoldfish opened the pull request:
@suspiciouscow @jamietheraccoon

jamesdooley4 commented:
Oh, and the pull request title doesn't seem to match the changes, as the PR is implementing the whole subsystem (not just changing capitalization).

bobtseattle commented on src/main/java/frc/team2412/robot/Subsystems.java:
```diff
@@ -35,7 +35,7 @@
         hardware = h;
         if (CLIMB_ENABLED)
-            climbSubsystem = new ClimbSubsystem(hardware.climbFixed1, hardware.climbFixed2, hardware.climbAngled1, hardware.climbAngled2, hardware.climbAngle);
+            climbSubsystem = new ClimbSubsystem(hardware.climbFixed1, hardware.climbFixed2, hardware.climbAngle);
```
This implies on the fixed arm has motors.  Consider correcting the name since one of the fixed motors is actually for angle.  The will reduce confusion and help with readability.

bobtseattle commented on src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:
```diff
@@ -1,12 +1,88 @@
-import com.ctre.phoenix.motorcontrol.can.WPI_TalonFX;
-
+import com.ctre.phoenix.motorcontrol.TalonFXControlMode;
+import com.ctre.phoenix.motorcontrol.can.TalonFX;
```
The leads recommend use of the WPI_TalonFX library, not TalonFX.  Sorry, if there was a miscommunication.
The WPI lib includes support for liveWindow and other features which will likely be very handy for debugging and troubleshooting the synchronization of the climbing subsystem.

bobtseattle commented on src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:
```diff
@@ -1,12 +1,88 @@
+
+    ClimbSubsystemState state;
+
+    public ClimbSubsystem(TalonFX climbFixed1, TalonFX climbFixed2, DoubleSolenoid climbAngle) {
```
Renaming these so they don't imply both are for the fixed arm would also be helpful for readability.

bobtseattle commented on src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:
```diff
@@ -1,12 +1,88 @@
+        climbMotor1 = climbFixed1;
+        climbMotor2 = climbFixed2;
+        solenoid = climbAngle;
+        setName("ClimbSubsystem");
```
Should it also set enabled to false on initialization?

bobtseattle commented on src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:
```diff
@@ -1,12 +1,88 @@
+        solenoid.set(extended ? DoubleSolenoid.Value.kForward : DoubleSolenoid.Value.kReverse);
+    }  
+    
+    public void extendFixedArm() {
```
Assuming a check for climbSystemEnabled will be implemented so the arms can't accidentally be extended before end game?

Also assuming there will be logic TBD to make sure the arms stop (retracting or extending) automatically at full extension or retraction so as to not damage the robot?

bobtseattle commented on src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:
```diff
@@ -1,12 +1,88 @@
+    }
+
+    public void extendAngledArm() {
+        climbMotor2.set(TalonFXControlMode.PercentOutput, ClimbConstants.MAX_SPEED);
```
Assuming a check for climbSystemEnabled will be implemented so the arms can't accidentally be extended before end game?

bobtseattle commented on src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:
```diff
@@ -1,12 +1,88 @@
-    public ClimbSubsystem(WPI_TalonFX climbFixed1, WPI_TalonFX climbFixed2, WPI_TalonFX climbAngled1, WPI_TalonFX climbAngled2, DoubleSolenoid climbAngle) {
 
+    public static class ClimbConstants {
+        public static final double MAX_SPEED = 1;
```
To reduce the chance of damage, you might consider a max speed other than 1 until we see how climb reacts to being powered up and arm being extended / retracted.

bobtseattle reviewed (changes requested):
Great start!

galaxygoldfish commented:
- Added soft limits to the motors in the constructor
- Added motor checks in periodic to make sure that they don't go past the limit (might remove later, just for safety?)
- Change motor variable names
- Change motor types to WPI_TalonFX
- Change motor speed to 0.5 for testing purposes for safety
- Check if subsystem is enabled before starting/changing motor state to prevent accidental climb
- Fix motor types and names in Hardware class to adapt to new constructor with two motors

galaxygoldfish commented:
- Added soft limits to the motors in the constructor
- Added motor checks in periodic to make sure that they don't go past the limit (might remove later, just for safety?)
- Change motor variable names
- Change motor types to WPI_TalonFX
- Change motor speed to 0.5 for testing purposes for safety
- Check if subsystem is enabled before starting/changing motor state to prevent accidental climb
- Fix motor types and names in Hardware class to adapt to new constructor with two motors

galaxygoldfish commented:
- Added soft limits to the motors in the constructor
- Added motor checks in periodic to make sure that they don't go past the limit (might remove later, just for safety?)
- Change motor variable names
- Change motor types to WPI_TalonFX
- Change motor speed to 0.5 for testing purposes for safety
- Check if subsystem is enabled before starting/changing motor state to prevent accidental climb
- Fix motor types and names in Hardware class to adapt to new constructor with two motors

galaxygoldfish commented on src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:
```diff
@@ -1,12 +1,88 @@
+
+     public void setDisabled() {
+        state = ClimbSubsystemState.DISABLED;
+    }
```
These are to avoid accidental climb, making sure that the drivers are aware that the robot is climbing

jamesdooley4 commented on src/main/java/frc/team2412/robot/Hardware.java:
```diff
@@ -1,5 +1,6 @@
 package frc.team2412.robot;
 
+import com.ctre.phoenix.motorcontrol.can.TalonFX;
```
Delete - this shouldn't be needed

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:
```diff
@@ -1,12 +1,107 @@
+        this.climbFixedMotor = climbFixedMotor;
+        this.climbDynamicMotor = climbDynamicMotor;
+        solenoid = climbAngle;
+        climbFixedMotor.configSupplyCurrentLimit(MOTOR_CURRENT_LIMIT);
+        climbDynamicMotor.configSupplyCurrentLimit(MOTOR_CURRENT_LIMIT);
+        climbFixedMotor.configForwardSoftLimitThreshold(MAX_ENCODER_TICKS);
+        climbFixedMotor.configReverseSoftLimitThreshold(MIN_ENCODER_TICKS);
+        climbFixedMotor.configForwardSoftLimitEnable(true, 0);
+        climbFixedMotor.configReverseSoftLimitEnable(true, 0);
+        climbDynamicMotor.configForwardSoftLimitThreshold(MAX_ENCODER_TICKS);
+        climbDynamicMotor.configReverseSoftLimitThreshold(MIN_ENCODER_TICKS);
+        climbDynamicMotor.configReverseSoftLimitEnable(true, 0);
+        climbDynamicMotor.configForwardSoftLimitEnable(true, 0);
```
Recommend to use TalonFXConfiguration() to set these all at once (and enable re-use across motors so it's more concise since several of the config settings are the same)

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:
```diff
@@ -1,12 +1,107 @@
+        }
+    }
+
+    public void setEnabled() {
+        state = ClimbSubsystemState.ENABLED;
+    }
+
+    public void setDisabled() {
+        state = ClimbSubsystemState.DISABLED;
+    }
```
What are these for?  They don't seem to reflect or change any physical state of the robot.

galaxygoldfish replied:
It's a state intended to make sure that the drivers don't accidentally cause the robot to climb, so the idea is that they would first press a button to get into climb mode, during endgame and then they would press the climb button or whatever we decide on doing for the climb automation

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:
```diff
@@ -1,12 +1,107 @@
+    
+    public void retractFixedArm() {
+        if (state == ClimbSubsystemState.ENABLED)
+            climbFixedMotor.set(-ClimbConstants.TEST_SPEED);
+    }
```
Consider having separate constants for extend & retract, as they may want to move at different speeds or with different power levels.

jamesdooley4 replied:
Ditto for the angledArm methods

galaxygoldfish commented:
- Now using TalonFXConfiguration
- Updated to use a faster speed for extending the arms and a slower speed to retract for safety, will change later during testing
